### ISSUE TITLE
Record<A,B> is assignable to Record<B,A> when A,B are tagged primitives

### ISSUE THREAD

**Issue by valkum** [microsoft/TypeScript#45796]
# Bug Report

When having two types of type Record with tagged primitives, one with a tagged primitive as a key and the other with a tagged primitive as a value, typescript 4.4 does not correctly complains about assignments

### 🔎 Search Terms
Record, tagged primitive, mapped types

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about mapped type

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDOwBOBLAdgcygMigG9UIBDAEwHtMAbEKAfQeFOwC4oBXTdAR0+hIQAWwBGlGgF8A3AChQkKACEEyNFlwFiZKrXpMW7Lj36CR4qXPnhoAJVW2IAY0qpyAHhgAaJQD5rigCSDs6uHkpeMP6yLpgoUADWECAAjKoA5ABmlJTpUKRIsHKx8UkgAEwZ2ZTleQVFMdTxAG6kNAJpiOmipKh1hUrFTcBQre0QlV09AF79ylYlI8AQKBz2iIQA2mUpALocYx1eUNvJ5fujbQLlMrKNcSNYY+jkHMGIyyhyQA)

### 💻 Code

```ts
type A = string & {readonly __tag: unique symbol};
type B = string & {readonly __tag: unique symbol};

type R = Record<A,B[]>
type I = Record<B,A>

const key1 = 'foo' as A;
const key2 = 'foo2' as A;
const value1 = 'bar' as B;
const value2 = 'baz' as B;

const test: R = {[key1]: [value1], [key2]: [value2]};

const invalid: I = test; // This should throw an error. When not for the simple types, by the fact that R has an Array as value and I a simple string

```

### 🙁 Actual behavior

Typescript compiles fine

### 🙂 Expected behavior

Throwing an error at the assignment

**Comment by andrewbranch** [comment]
There’s nothing inconsistent about this: consider an interface that has both index signatures:

```ts
interface Both {
  [key: A]: B[];
  [key: B]: A;
}
```

Put an `A` in, get a `B[]` out. Put a `B` in, get an `A` out. An `R` and an `I` are assignable to each other in either direction because their index signatures don’t overlap. In other words, assigning `test` to `invalid` doesn’t break the contract of `I` because the type of `test` says nothing about keys of type `B`, which is the only thing that `I` cares about.

**Comment by Jamesernator** [comment]
If you want this to work the way you expect you need to actually pick a unique symbol for each.

i.e.:

```ts
const aBrand = Symbol();
const bBrand = Symbol();
type A = string & {readonly __tag: typeof aBrand};
type B = string & {readonly __tag: typeof bBrand};
```

**Comment by Jamesernator** [comment]
Actually nevermind I understand @andrewbranch's example now, it does seem kinda weird still though. Is there a way to declare that extra keys shouldn't be present in TypeScript?

**Comment by fatcerberus** [comment]
No, TypeScript always allows extra properties as long as they don’t conflict with anything in the type definition (except for excess property checks in contextually-typed object literals).  See #12936.
